include/rmm/device_buffer.hpp

```diff
@@ -392,4 +392,26 @@ class device_buffer {
       mr::get_default_resource()};  ///< The memory resource used to
                                     ///< allocate/deallocate device memory
 };
+
+/**--------------------------------------------------------------------------*
+ * @brief Copies rmm::device_buffer to a preallocated host buffer.
+ *
+ * Copies device memory asynchronously on the specified stream
+ *-------------------------------------------------------------------------**/
+void copy_to_host(const device_buffer& db, void* hb, cudaStream_t stream = 0) {
+  if (hb == nullptr) {
+    throw std::runtime_error{"Cannot copy to `nullptr`."};
+  }
+  cudaError_t err = cudaMemcpyAsync(hb,
+                                    db.data(),
+                                    db.size(),
+                                    cudaMemcpyDeviceToHost,
+                                    stream);
+  if (err != cudaSuccess) {
+    throw std::runtime_error{"Failed to copy to host."};
+  }
+  if (cudaSuccess != cudaStreamSynchronize(stream)) {
```

Comment:
This function is supposed to be asynchronous.

Reply:
Making it asynchronous causes it to be harder to use from Python. Though we could call `cudaStreamSynchronize` in Python if that would be preferred.

Reply:
Any function that takes a stream should be asynchronous, otherwise that defeats the purpose of providing a stream. It's then the user's responsibility to respect stream ordering or do the synchronization.

Reply:
Any objection to getting rid of the stream then?

Reply:
Every other device buffer function that does an allocation or copy has a stream argument. I don't like the idea of deviating from that.

Reply:
Makes sense. Have moved this to the Python layer.

Reply:
Out of curiosity, is the stream exposed at the Python layer? I assume at the Python level the default stream would always be used, in which case a stream sync isn't necessary.

Reply:
It probably makes sense to allow users to set the stream in Python (CuPy allows this for instance), but fallback to the default stream if unspecified. That's how the recent commits are handling this.

Reply:
How is a stream represented in Python? From what I recall, numerous CUDA Python libraries all have their own abstraction for a stream, so I'm curious what we're using.

Reply:
Thus far we have been handling pointers like this in Python by turning them into Python `int`s. We then convert them to `uintptr_t` in Cython before casting them to whatever C/C++ needs them as. That's what I've done here as well.